src/Type/ArrayType.php

```diff
@@ -659,6 +659,24 @@ public function traverse(callable $cb): Type
 		return $this;
 	}
 
+	public function traverseWithVariance(TemplateTypeVariance $variance, callable $cb): Type
```

Comment:
`traverseWithVariance` seems like a combination of `getReferencedTemplateTypes` + `traverse`. Could we somehow use that instead of adding a new `traverseWithVariance` method on `Type`?

Reply:
And even if not, it might be possible to somehow deduplicate the logic between `traverse` and `traverseWithVariance`? It might be hard to keep them in sync.

Reply:
Like maybe `traverse` could just call `traverseWithVariance` with a little bit different callback :)

Reply:
> `traverseWithVariance` seems like a combination of `getReferencedTemplateTypes` + `traverse`. Could we somehow use that instead of adding a new `traverseWithVariance` method on `Type`?

Funny you mention that. In one of my early prototypes, I've had it implemented that way. But it felt too fragile because it relied on object identity (`TemplateTypeReference`'s template type === template type passed into the `traverse` callback), so I threw it away...

> Like maybe traverse could just call traverseWithVariance with a little bit different callback :)

Hmm, could be, I'll see where it leads.

Reply:
I've added the prototype code in jiripudil/phpstan-src#7, I'm curious if you think that could be a way. I have a bad feeling about it because of the reliance on object identity, but it completely eliminates the issue with logic duplication, and it works 🤷

Reply:
I like that, can you please simplify this #2485 PR with this approach? :)

And BTW in case `===` would not work for us in the future, we could use `Type::equals()` right?

Reply:
Oh, that's the tricky bit. Consider this example:

```php
/** @template T */
class Foo {
    /**
     * @param callable(T): T $cb
     */
    public function doFoo(callable $cb): void { /* ... */ }
}

/** @param Foo<*> */
function bar(Foo $foo) {
    $foo->doFoo(/* ... */);
}
```

Now, at some point in the resolution, PHPStan needs to replace the template types referenced in the type `callable(T): T`. The thing is that it's the same template type which happens to be in two positions with different variance, and we need to distinguish between the two occurrences and treat each of them differently.

This simple approach uses object identity for this, relying solely on the fact that PHPStan creates a new instance of the template type for each of the occurrences. The two template types are otherwise equal in all regards, therefore, no, we need to stick with `===`, and the code will only work as long as this prerequisite remains fulfilled.

I like the simplicity of this solution, but I can't shake off the feeling that it's relying on an implementation detail and that it may all break as soon as somebody comes with an optimization of how types are created.

If you're ok with it, I am too, but I want to make sure you have all the relevant information to make the decision :)

Reply:
Yes, it's okay now and if we need it, we have a solution for that (a new method on Type).

This would make a really nice comment next to the `===` operator 😊

Reply:
Done and done :)